### DISCUSSION TITLE
modbus tcp register und ID problem

### DISCUSSION THREAD

**Discussion by MiRoGa-NV** [tbnobody/OpenDTU#2134] [Q&A]
Hallo,

wir haben heute versucht mit meinem Speicherhertseller herauszufinden auf welchem Register die PV-Leistung Total für meine 4 WR von Hoymiles (HM-1500 4T) abgeholt wird. Es scheint Adresse 40192 zu sein.  Meine beiden Speicher sind von Sax Heimspeicher. Ich möchte die PV_Leistung gesamt auf der Web-App, sowie auf dem E-Paper Display sehen.

<img width="1433" alt="Bildschirmfoto 2024-07-09 um 17 42 24" src="https://github.com/tbnobody/OpenDTU/assets/138251017/e2171053-f080-46e2-9934-464a230edf05">

![tempImagewNqEAz](https://github.com/tbnobody/OpenDTU/assets/138251017/b99914c7-2ffe-4032-866e-10d797c10422)

Folgendes haben wir heute eingestellt:

![image001](https://github.com/tbnobody/OpenDTU/assets/138251017/426acc66-87ee-4dcd-b857-01afb0c89907)

Versucht haben wir:
Erhöhung des Scale Faktors auf 100, kein Erfolg, aber 0W Anzeige auf dem Display
alte Gateway Software SAX mit 2Byte Länge, kein Erfolg, aber 0W Anzeige auf dem Display
Gateway Update SAX, 2 Byte Länge statt 4 Byte, gar keine W Anzeige mehr auf dem Display
Versucht über Stammregister SM Power und Einbindung aller P Inverter, kein Erfolg, aber 0W Anzeige auf dem Display
 
Das alles über TCP Protokoll!
 
Punkte die noch offen sind:
DTU evtl. auf eigene Scale-Faktor Registeradresse prüfen und evtl. eintragen und versuchen
evtl. vergibt die eigene DTU eine eigene Unit ID anstelle von 125.

OpenDTU wird von einem Modbus-Server erweitert. #1893
Der Modbus-Server bedient TCP an Port 502.
Bei Modbus ID 1 ahmt der Server die Modbus-Register im ursprünglichen DTUPro nach.
Bei Modbus ID 125 bedient der Server einen SunSpec-kompatiblen "Total Inverter", der
stellt die aggregierten OpenDTU-Daten aller registrierten Wechselrichter zur Verfügung.
Bei Modbus ID 243 bedient der Server ein SunSpec-Messgerät, das Wechselstrom und Wechselstromausbeute liefert, als würde er alle registrierten Wechselrichter messen. Derselbe Ansatz wie beim "Gesamtwechselrichter", aber jetzt als Zähler verkleidet.

Kann hier jemand helfen, das Problem zu lösen?

**Comment by b0661** [comment]
Hallo @MiRoGa-NV ,

das Image aus dem pull request #1893 liefert an Adresse 40192 einen float32 Wert (SunSpec Model 113 - Inverter (Three Phase) FLOAT Model). 

Ich könnte mir diese Probleme vorstellen:

1) Die Heimspeicher-SW erwartet einen int32 Wert. Sieht von den Einstellungen her so aus.
2) Keine Verbindung der Heimspeicher-SW zum OpenDTU Modbus Port - Modbus ist nicht aktiviert. Darauf könnte die 0W Anzeige hindeuten.
3) Keine Verbindung der Heimspeicher-SW zum OpenDTU Modbus Port - die OpenDTU ist mit dem "Standard"-Image geflashed. Darauf könnte die 0W Anzeige hindeuten.
4) Keine Verbindung der Heimspeicher-SW zur OpenDTU - Netzwerkprobleme. Darauf könnte die 0W Anzeige auch hindeuten.

Lösungen:

1) In der Heimspeicher-SW die Abfrage des Wertes auf float32 setzen. Falls dies nicht möglich ist müsste ich die Modbus-SW um das entsprechende SunSpec INT Model erweitern. Sollte relativ einfach möglich sein.
2) Modbus unter Einstellungen aktivieren
3) Das Image aus dem pull request #1893 bauen und flashen.
4) Evtl. Heimspeicher-Fehlermeldungen prüfen.

Bobby

**Reply by MiRoGa-NV** [reply]
Hallo Bobby,

danke für die schnelle Antwort.

hier meine Systeminformation:

<img width="678" alt="Bildschirmfoto 2024-07-10 um 09 33 59" src="https://github.com/tbnobody/OpenDTU/assets/138251017/f62329e9-5d20-4a1b-bacf-e8ecf4e6b7c5">

<img width="678" alt="Bildschirmfoto 2024-07-10 um 09 34 11" src="https://github.com/tbnobody/OpenDTU/assets/138251017/4525c6c7-2a8c-4cb7-822a-cc86dae3582f">

In HA und Evcc ist mit tcpip über port 502 die PV-Leistung sichtbar.
Vermutung ist das wir die falsche ID des openDTU einsetzen. Wie ich gelesen habe sind ID: 1, ID: 125 und ID: 243 möglich.


Screenshot von evccum 09.41 von heute:

<img width="1124" alt="Bildschirmfoto 2024-07-10 um 09 41 22" src="https://github.com/tbnobody/OpenDTU/assets/138251017/fa6dadba-c283-4059-a46b-7796f53f247d">


Gruß
Michael

**Reply by b0661** [reply]
Hallo Michael,

> In HA und Evcc ist mit tcpip über port 502 die PV-Leistung sichtbar.
> Vermutung ist das wir die falsche ID des openDTU einsetzen. Wie ich gelesen habe sind ID: 1, ID: 125 und ID: 243 möglich.

ID 125 stimmt, allerdings liefert Adresse 40192 einen 32bit Fließkommawert.

Zum Testen kannst Du z.B. Adresse 40170 auslesen (int16, __KEIN__ Haken by 4 bytes, Factor 1). Als Wert sollte 113 erscheinen.

Falls das nicht funktioniert mit Adresse 40171 versuchen. Evtl. hat der HeimspeicherSW-Ersteller eine andere Vorstellung von der Addresszählung. 

Falls das auch nicht funktioniert zeig mal bitte deine Modbus-Einstellungen. Nachdem mehrere Stellen auf Port 502 zugreifen ist es wichtig dass die "Maximale Anzahl der Modbus TCP-Clients" enstprechend hochgesetzt ist.


Bobby

**Reply by MiRoGa-NV** [reply]
Hallo Bobby,

du schreibst:
`Modbus unter Einstellungen aktivieren`

in meiner Version der openDTU kann ich keinen Reiter Modbus in den Einstellungen finden.
Firmware : 24.5.6

wird wohl das Problem sein, dass in dieser Version der Modbus nicht enthalten ist.

Gruß
Michael

**Reply by stefan123t** [reply]
> 3. Das Image aus dem pull request https://github.com/tbnobody/OpenDTU/pull/1893 bauen und flashen.

@MiRoGa-NV ich vermute der Schritt mit dem Build des PR [pr_modbus](https://github.com/b0661/OpenDTU/tree/pr_modbus) / #1893 steht bei Dir noch aus, oder ?

**Reply by MiRoGa-NV** [reply]
Ja, so ist es.
Ich muss erst verstehen wie das funktioniert und lesen bevor ich etwas falsch mache.
Meine DTU ist eine DX-DTU-MAX für HM mit ES32.

**Reply by MiRoGa-NV** [reply]
Welche der x.bin muss ich flashen, es stehen mehrere zur Verfügung. Und was ist der Unterschied.

<img width="178" alt="Bildschirmfoto 2024-07-11 um 14 52 50" src="https://github.com/tbnobody/OpenDTU/assets/138251017/992e8a76-9e53-46e6-aa1d-cfa5705bcca6">

und kann ich dann zur meiner alten Version zurückspringen, wenn es nicht funktioniert?

Michael

**Reply by b0661** [reply]
firmware.bin

**Reply by MiRoGa-NV** [reply]
wie kann ich jetzt prüfen ob ich die richtigen Einstellungen habe.
DTU ist ist mit neuer Firmware geflasht.

**Reply by b0661** [reply]
Unter Einstellungen Modbus aktivieren. Unter Info kannst Du die Einstellungen prüfen. Nach dem Aktivieren sollten die Daten per Modbus verfügbar sein

**Reply by MiRoGa-NV** [reply]
Ich habe keine sichtbaren Einstellungen zur Modbus Aktivierung.

<img width="838" alt="Bildschirmfoto 2024-07-12 um 08 04 34" src="https://github.com/user-attachments/assets/e9e673d4-02d3-44ba-805b-dcb7086b5204">

<img width="370" alt="Bildschirmfoto 2024-07-12 um 08 05 26" src="https://github.com/user-attachments/assets/8893b1cb-911a-4251-ba5f-7971ebf90550">

jedoch beim nochmaligen Kompilieren folgenden Fehler/Warnung: Ist mir gestern nicht aufgefallen, erst heute Morgen.

<img width="1404" alt="Bildschirmfoto 2024-07-12 um 08 07 12" src="https://github.com/user-attachments/assets/9898b349-6ee1-4989-af55-23cdb51a528d">

**Comment by b0661** [comment]
Hallo Michael,

Du hast anscheinend nicht das Image zum Pull request erzeugt. 

Das sollte sichtbar sein:
![Screenshot_20240712_094511](https://github.com/user-attachments/assets/d1edba55-4150-437b-b768-25430e342cf2)
![Screenshot_20240712_094542](https://github.com/user-attachments/assets/9f50d792-3816-4adb-a9d5-bfdef5e3986f)

Hier ist ein Image für generic_esp32:
[firmware.bin.gz](https://github.com/user-attachments/files/16190516/firmware.bin.gz)

Vor dem flashen musst Du das Image noch entpacken:

```
  gunzip firmware.bin.gz
```

Falls Dir die Verwendung des Images zu risikoreich ist musst Du den Pull Request von github abholen (siehe https://docs.github.com/de/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally).

```
  cd .../OpenDTU
  git fetch origin pull/1893/head:pr_modbus
  git switch pr_modbus
```

Anschließend wie gehabt bauen und flashen.

VG Bobby

**Reply by MiRoGa-NV** [reply]
Danke,  ich hab dein Image benutzt und habe den Modbus aktiviert.

<img width="237" alt="Bildschirmfoto 2024-07-13 um 17 47 39" src="https://github.com/user-attachments/assets/b6890cb3-6d10-4a59-a3f9-2f36e97b499e">

Wenn ich jetzt versuche Adresse 40192 auszulesen erhalte ich folgendes:

<img width="755" alt="Bildschirmfoto 2024-07-13 um 19 08 53" src="https://github.com/user-attachments/assets/de23ce6b-51f5-4f82-b231-b9bb2526b610">

Dieser Wert sollte eigentlich ausgegeben werden.

<img width="339" alt="Bildschirmfoto 2024-07-13 um 19 24 34" src="https://github.com/user-attachments/assets/aac408e4-bc95-4fa6-8e33-b408fbb19008">


Um zu prüfen ob ich einen Fehler verursacht habe, habe ich alles auch mit dem Speicher durchgeführt.

Diese Adressen gibt Sax im Handbuch an.

<img width="781" alt="Bildschirmfoto 2024-07-13 um 19 15 54" src="https://github.com/user-attachments/assets/8009b42a-dfd5-43bc-9302-0dd466ea1a03">

Lese ich den  SOC  aktuell 54% vom Speicher mit 0x2e so erhalte ich:


<img width="764" alt="Bildschirmfoto 2024-07-13 um 19 01 45" src="https://github.com/user-attachments/assets/fc243be9-50c6-41bd-b951-25132a09842b">

Wo, ist der Fehler? Ist es möglich, dass die OpenDTU mit Modbus am Ethernet hängen muß, wie die Originale von HM? Der Speicher ist über Ethernet verbunden.

VG Michael

**Reply by b0661** [reply]
Hallo,

versuch mal die scan rate auf 5000 ms zu setzen. 1 Sekunde ist eventuell zu schnell.

**Reply by MiRoGa-NV** [reply]
Hab sogar auf 10 Sekunden erhöht, meldet immer noch Illegale Data Address.

**Reply by b0661** [reply]
Schau Dir bitte mal die Konsole der OpenDTU an (unter Info). Evtl. gibt es da mehr Info.

**Reply by MiRoGa-NV** [reply]
gesagt getan

```
20:29:57.037 > Fetch inverter: 116183134121
20:29:57.078 > TX RealTimeRunData Channel: 40 --> 15 83 13 41 21 80 19 86 00 80 0B 00 66 92 C7 A5 00 00 00 00 00 00 00 00 49 AC 02 
20:29:57.093 > Interrupt received
20:29:57.102 > RX Channel: 3 --> 95 83 13 41 21 83 13 41 21 01 00 01 01 0F 00 13 00 16 00 34 00 3B 00 00 86 43 54 | -80 dBm
20:29:57.258 > Interrupt received
20:29:57.273 > RX Channel: 3 --> 95 83 13 41 21 83 13 41 21 84 13 86 00 D6 00 D8 00 09 02 C1 00 FC 00 02 D6 AF C7 | -80 dBm
20:29:57.651 > RX Period End
20:29:57.651 > Middle missing
20:29:57.651 > Request retransmit: 2
20:29:57.651 > TX RequestFrame Channel: 61 --> 15 83 13 41 21 80 19 86 00 82 78 
20:29:57.745 > RX Period End
20:29:57.745 > Middle missing
20:29:57.745 > Request retransmit: 2
20:29:57.745 > TX RequestFrame Channel: 75 --> 15 83 13 41 21 80 19 86 00 82 78 
20:29:57.756 > Interrupt received
20:29:57.765 > RX Channel: 23 --> 95 83 13 41 21 83 13 41 21 02 00 00 89 DD 03 92 03 B4 01 0F 00 14 00 16 00 36 DF | -80 dBm
20:29:57.882 > RX Period End
20:29:57.883 > Middle missing
20:29:57.883 > Request retransmit: 3
20:29:57.883 > TX RequestFrame Channel: 3 --> 15 83 13 41 21 80 19 86 00 83 79 
20:29:58.011 > RX Period End
20:29:58.011 > Middle missing
20:29:58.011 > Request retransmit: 3
20:29:58.011 > TX RequestFrame Channel: 23 --> 15 83 13 41 21 80 19 86 00 83 79 
20:29:58.143 > RX Period End
20:29:58.143 > Middle missing
20:29:58.143 > Request retransmit: 3
20:29:58.143 > TX RequestFrame Channel: 40 --> 15 83 13 41 21 80 19 86 00 83 79 
20:29:58.255 > RX Period End
20:29:58.255 > Middle missing
20:29:58.255 > Retransmit timeout
20:30:02.054 > Fetch inverter: 116183132669
20:30:02.054 > Request SystemConfigPara
20:30:02.081 > TX RealTimeRunData Channel: 61 --> 15 83 13 26 69 80 19 86 00 80 0B 00 66 92 C7 AA 00 00 00 00 00 00 00 00 B9 ED 93 
20:30:02.118 > Interrupt received
20:30:02.126 > RX Channel: 23 --> 95 83 13 26 69 83 13 26 69 02 00 00 7D 74 03 75 03 AA 01 10 00 11 00 15 00 2D 79 | -80 dBm
20:30:02.211 > Interrupt received
20:30:02.223 > RX Channel: 75 --> 95 83 13 26 69 83 13 26 69 03 00 38 00 00 79 9A 00 00 80 95 03 3D 03 A1 09 3C F1 | -80 dBm
20:30:02.613 > RX Period End
20:30:02.613 > Last missing
20:30:02.613 > Request retransmit: 4
20:30:02.613 > TX RequestFrame Channel: 75 --> 15 83 13 26 69 80 19 86 00 84 51 
20:30:02.743 > RX Period End
20:30:02.743 > Last missing
20:30:02.743 > Request retransmit: 4
20:30:02.743 > TX RequestFrame Channel: 3 --> 15 83 13 26 69 80 19 86 00 84 51 
20:30:02.975 > RX Period End
20:30:02.975 > Last missing
20:30:02.975 > Request retransmit: 4
20:30:02.975 > TX RequestFrame Channel: 23 --> 15 83 13 26 69 80 19 86 00 84 51 
20:30:03.109 > RX Period End
20:30:03.109 > Last missing
20:30:03.109 > Request retransmit: 4
20:30:03.109 > TX RequestFrame Channel: 40 --> 15 83 13 26 69 80 19 86 00 84 51 
20:30:03.120 > Interrupt received
20:30:03.133 > RX Channel: 75 --> 95 83 13 26 69 83 13 26 69 84 13 86 00 CD 00 DA 00 09 02 AE 00 F3 00 01 5E 3C A6 | -80 dBm
20:30:03.141 > RX Period End
20:30:03.141 > Middle missing
20:30:03.141 > Request retransmit: 1
20:30:03.141 > TX RequestFrame Channel: 61 --> 15 83 13 26 69 80 19 86 00 81 54 
20:30:03.156 > Interrupt received
20:30:03.168 > RX Channel: 3 --> 95 83 13 26 69 83 13 26 69 01 00 01 01 11 00 14 00 16 00 36 00 3C 00 00 7D 47 B7 | -80 dBm
20:30:03.257 > RX Period End
20:30:03.257 > Success
20:30:03.301 > TX SystemConfigPara Channel: 75 --> 15 83 13 26 69 80 19 86 00 80 05 00 66 92 C7 AA 00 00 00 00 00 00 00 00 77 E2 5C 
20:30:03.317 > Interrupt received
20:30:03.324 > RX Channel: 61 --> 95 83 13 26 69 83 13 26 69 81 00 01 03 E8 00 00 03 E8 FF FF FF FF 01 68 7D F8 F9 | -80 dBm
20:30:03.506 > RX Period End
20:30:03.506 > Success
20:30:07.072 > Fetch inverter: 116183133287
20:30:07.081 > TX RealTimeRunData Channel: 3 --> 15 83 13 32 87 80 19 86 00 80 0B 00 66 92 C7 AF 00 00 00 00 00 00 00 00 E9 D2 03 
20:30:07.688 > RX Period End
20:30:07.688 > All missing
20:30:07.688 > Nothing received, resend whole request
20:30:07.688 > TX RealTimeRunData Channel: 23 --> 15 83 13 32 87 80 19 86 00 80 0B 00 66 92 C7 AF 00 00 00 00 00 00 00 00 E9 D2 03 
20:30:08.199 > RX Period End
20:30:08.199 > All missing
20:30:08.199 > Nothing received, resend whole request
20:30:08.199 > TX RealTimeRunData Channel: 40 --> 15 83 13 32 87 80 19 86 00 80 0B 00 66 92 C7 AF 00 00 00 00 00 00 00 00 E9 D2 03 
20:30:08.240 > Interrupt received
20:30:08.248 > RX Channel: 75 --> 95 83 13 32 87 83 13 32 87 02 00 00 7E 79 03 39 03 92 01 10 00 10 00 14 00 2B 05 | -80 dBm
20:30:08.316 > Interrupt received
20:30:08.326 > RX Channel: 40 --> 95 83 13 32 87 83 13 32 87 03 00 36 00 00 76 5C 00 00 7E AC 03 32 03 95 09 30 C6 | -80 dBm
20:30:08.722 > RX Period End
20:30:08.722 > Last missing
20:30:08.722 > Request retransmit: 4
20:30:08.722 > TX RequestFrame Channel: 61 --> 15 83 13 32 87 80 19 86 00 84 AB 
20:30:08.851 > RX Period End
20:30:08.851 > Last missing
20:30:08.851 > Request retransmit: 4
20:30:08.851 > TX RequestFrame Channel: 75 --> 15 83 13 32 87 80 19 86 00 84 AB 
20:30:08.984 > RX Period End
20:30:08.984 > Last missing
20:30:08.984 > Request retransmit: 4
20:30:08.984 > TX RequestFrame Channel: 3 --> 15 83 13 32 87 80 19 86 00 84 AB 
20:30:09.117 > RX Period End
20:30:09.117 > Last missing
20:30:09.117 > Request retransmit: 4
20:30:09.117 > TX RequestFrame Channel: 23 --> 15 83 13 32 87 80 19 86 00 84 AB 
20:30:09.128 > Interrupt received
20:30:09.136 > RX Channel: 75 --> 95 83 13 32 87 83 13 32 87 84 13 86 00 BA 00 D7 00 08 02 8E 00 F8 00 04 7B C6 2C | -80 dBm
20:30:09.266 > RX Period End
20:30:09.266 > Middle missing
20:30:09.266 > Request retransmit: 1
20:30:09.266 > TX RequestFrame Channel: 40 --> 15 83 13 32 87 80 19 86 00 81 AE 
20:30:09.381 > RX Period End
20:30:09.381 > Middle missing
20:30:09.381 > Retransmit timeout
20:30:12.090 > Fetch inverter: 116183133420
20:30:12.122 > TX RealTimeRunData Channel: 61 --> 15 83 13 34 20 80 19 86 00 80 0B 00 66 92 C7 B4 00 00 00 00 00 00 00 00 19 6C F7 
20:30:12.135 > Interrupt received
20:30:12.143 > RX Channel: 3 --> 95 83 13 34 20 83 13 34 20 01 00 01 01 10 00 11 00 14 00 2E 00 37 00 00 7F D0 37 | -80 dBm
20:30:12.152 > Interrupt received
20:30:12.161 > RX Channel: 3 --> 95 83 13 34 20 83 13 34 20 02 00 00 87 EF 03 37 03 8E 01 11 00 12 00 13 00 30 67 | -80 dBm
20:30:12.704 > RX Period End
20:30:12.704 > Last missing
20:30:12.704 > Request retransmit: 3
20:30:12.704 > TX RequestFrame Channel: 75 --> 15 83 13 34 20 80 19 86 00 83 0D 
20:30:12.839 > Interrupt received
20:30:12.851 > RX Channel: 3 --> 95 83 13 34 20 83 13 34 20 03 00 33 00 00 82 CE 00 00 81 D9 03 4F 03 64 09 32 A1 | -80 dBm
20:30:12.857 > RX Period End
20:30:12.857 > Last missing
20:30:12.857 > Request retransmit: 4
20:30:12.857 > TX RequestFrame Channel: 3 --> 15 83 13 34 20 80 19 86 00 84 0A 
20:30:12.883 > RX Period End
20:30:12.883 > Last missing
20:30:12.883 > Request retransmit: 4
20:30:12.883 > TX RequestFrame Channel: 23 --> 15 83 13 34 20 80 19 86 00 84 0A 
20:30:12.907 > Interrupt received
20:30:12.917 > RX Channel: 61 --> 95 83 13 34 20 83 13 34 20 84 13 85 00 BE 00 D8 00 08 02 95 00 DD 00 02 98 E5 DC | -80 dBm
20:30:12.998 > RX Period End
20:30:12.998 > Success
20:30:17.064 > Fetch inverter: 116183134121
20:30:17.088 > TX RealTimeRunData Channel: 40 --> 15 83 13 41 21 80 19 86 00 80 0B 00 66 92 C7 B9 00 00 00 00 00 00 00 00 89 34 46 
20:30:17.137 > Interrupt received
20:30:17.149 > RX Channel: 75 --> 95 83 13 41 21 83 13 41 21 02 00 00 89 DD 03 92 03 B4 01 10 00 15 00 17 00 39 CF | -80 dBm
20:30:17.179 > Interrupt received
20:30:17.188 > RX Channel: 75 --> 95 83 13 41 21 83 13 41 21 03 00 3F 00 00 88 56 00 00 88 F4 03 B5 03 B9 09 31 3F | -80 dBm
20:30:17.239 > Interrupt received
20:30:17.250 > RX Channel: 61 --> 95 83 13 41 21 83 13 41 21 84 13 85 00 E0 00 D8 00 0A 02 D1 00 FD 00 02 DC 59 1C | -80 dBm
20:30:17.721 > RX Period End
20:30:17.721 > Middle missing
20:30:17.721 > Request retransmit: 1
20:30:17.721 > TX RequestFrame Channel: 61 --> 15 83 13 41 21 80 19 86 00 81 7B 
20:30:17.751 > RX Period End
20:30:17.751 > Middle missing
20:30:17.751 > Request retransmit: 1
20:30:17.751 > TX RequestFrame Channel: 75 --> 15 83 13 41 21 80 19 86 00 81 7B 
20:30:17.885 > RX Period End
20:30:17.885 > Middle missing
20:30:17.885 > Request retransmit: 1
20:30:17.885 > TX RequestFrame Channel: 3 --> 15 83 13 41 21 80 19 86 00 81 7B 
20:30:18.020 > RX Period End
20:30:18.020 > Middle missing
20:30:18.020 > Request retransmit: 1
20:30:18.020 > TX RequestFrame Channel: 23 --> 15 83 13 41 21 80 19 86 00 81 7B 
20:30:18.167 > RX Period End
20:30:18.167 > Middle missing
20:30:18.167 > Request retransmit: 1
20:30:18.167 > TX RequestFrame Channel: 40 --> 15 83 13 41 21 80 19 86 00 81 7B 
20:30:18.182 > Interrupt received
20:30:18.189 > RX Channel: 75 --> 95 83 13 41 21 83 13 41 21 01 00 01 01 11 00 14 00 16 00 36 00 3D 00 00 86 43 49 | -80 dBm
20:30:18.264 > RX Period End
20:30:18.265 > Success
20:30:22.124 > Fetch inverter: 116183132669
20:30:22.260 > TX RealTimeRunData Channel: 61 --> 15 83 13 26 69 80 19 86 00 80 0B 00 66 92 C7 BE 00 00 00 00 00 00 00 00 B9 12 78 
20:30:22.269 > Interrupt received
20:30:22.277 > RX Channel: 23 --> 95 83 13 26 69 83 13 26 69 01 00 01 01 12 00 15 00 17 00 39 00 3E 00 00 7D 47 B9 | -80 dBm
20:30:22.289 > Interrupt received
20:30:22.298 > RX Channel: 3 --> 95 83 13 26 69 83 13 26 69 02 00 00 7D 74 03 75 03 AA 01 11 00 12 00 16 00 30 65 | -80 dBm
20:30:22.304 > Interrupt received
20:30:22.315 > RX Channel: 75 --> 95 83 13 26 69 83 13 26 69 03 00 3B 00 00 79 9A 00 00 80 95 03 3D 03 A1 09 3C F2 | -80 dBm
20:30:22.320 > Interrupt received
20:30:22.333 > RX Channel: 3 --> 95 83 13 26 69 83 13 26 69 84 13 85 00 D8 00 DA 00 09 02 C0 00 F3 00 01 F1 84 C9 | -80 dBm
20:30:22.638 > RX Period End
20:30:22.638 > Success
20:30:27.045 > Fetch inverter: 116183133287
20:30:27.091 > TX RealTimeRunData Channel: 75 --> 15 83 13 32 87 80 19 86 00 80 0B 00 66 92 C7 C3 00 00 00 00 00 00 00 00 EB 2F 90 
20:30:27.100 > Interrupt received
20:30:27.108 > RX Channel: 23 --> 95 83 13 32 87 83 13 32 87 01 00 01 01 13 00 11 00 15 00 2F 00 39 00 00 77 DD 3F | -80 dBm
20:30:27.270 > Interrupt received
20:30:27.280 > RX Channel: 61 --> 95 83 13 32 87 83 13 32 87 84 13 85 00 C4 00 D7 00 08 02 A2 00 F9 00 04 0D 51 9D | -80 dBm
20:30:27.654 > RX Period End
20:30:27.654 > Middle missing
20:30:27.654 > Request retransmit: 2
20:30:27.654 > TX RequestFrame Channel: 3 --> 15 83 13 32 87 80 19 86 00 82 AD 
20:30:27.756 > RX Period End
20:30:27.756 > Middle missing
20:30:27.756 > Request retransmit: 2
20:30:27.756 > TX RequestFrame Channel: 23 --> 15 83 13 32 87 80 19 86 00 82 AD 
20:30:27.888 > RX Period End
20:30:27.888 > Middle missing
20:30:27.888 > Request retransmit: 2
20:30:27.888 > TX RequestFrame Channel: 40 --> 15 83 13 32 87 80 19 86 00 82 AD 
20:30:27.899 > Interrupt received
20:30:27.905 > RX Channel: 75 --> 95 83 13 32 87 83 13 32 87 02 00 00 7E 79 03 39 03 92 01 11 00 11 00 15 00 2E 01 | -80 dBm
20:30:28.021 > RX Period End
20:30:28.021 > Middle missing
20:30:28.021 > Request retransmit: 3
20:30:28.021 > TX RequestFrame Channel: 61 --> 15 83 13 32 87 80 19 86 00 83 AC 
20:30:28.032 > Interrupt received
20:30:28.040 > RX Channel: 75 --> 95 83 13 32 87 83 13 32 87 03 00 39 00 00 76 5C 00 00 7E AC 03 32 03 95 09 30 C9 | -80 dBm
20:30:28.133 > RX Period End
20:30:28.133 > Success

```

**Reply by b0661** [reply]
Addressfehler werden nur als Warnungen betrachtet und die sind momentan im Image nicht aktiviert. Ich schau dass ich ein entsprechendes Image baue.

**Reply by b0661** [reply]
Hier ist ein generic_esp32 Image in dem das Logging für Modbus Warnungen und Fehler eingeschalten ist. Vielleicht gib es da ja mehr informationen.
[firmware.debug.bin.gz](https://github.com/user-attachments/files/16203324/firmware.debug.bin.gz)

**Reply by b0661** [reply]
Edit:
Die Adresse 40192 wie sie hier #1893 angegeben wird ist eine Adresse für die Home Assistant Modbus Integration. Sie ist absolut zu lesen - enthält also __nicht__ den oft in der Dokumentation verwendeten Holding Register Offset von 40001. 

Die Anfrage sollte so aussehen:
![Screenshot_20240713_205555](https://github.com/user-attachments/assets/35b7c9e1-8af0-4074-8b9c-b067c863f09e)

Erzeugt mit https://npulse.net/en/online-modbus.

**Reply by MiRoGa-NV** [reply]
das wird angefragt: `7D 03 00 BF 00 02`

<img width="598" alt="Bildschirmfoto 2024-07-13 um 22 36 39" src="https://github.com/user-attachments/assets/4bdbd37d-3f33-4ca9-b5be-b018baf2923c">


und das erhalte ich zurück `7D 83 02`
<img width="581" alt="Bildschirmfoto 2024-07-13 um 22 30 33" src="https://github.com/user-attachments/assets/8b6b9f53-8588-4763-b9a0-8affd808afff">

wenn ich das richtig sehe, wird bei dir ein CRC Wert (FE:13) übermittelt, dieser fehlt bei mir.

**Reply by b0661** [reply]
Hallo,

mir war die Adressangabe für Home Assistant Modbus selbst nicht mehr ganz klar. In https://github.com/tbnobody/OpenDTU/discussions/2134#discussioncomment-10040357 habe ich das jetzt korrigiert.

Du solltest also einfach die Adresse 40192 verwenden (und nicht 40001 abziehen).

VG Bobby

**Comment by astrakaro** [comment]
Hallo, ich komme mit dem Pull Request auch nicht klar und benötige die Firmware für die Freischaltung der Modbus Funktionalität. 
Könnte mir hier jemand weiterhelfen?

![chip_opendtu](https://github.com/user-attachments/assets/8ef90ae6-0f88-47cf-bd75-c44c6c181d06)

Vielen Dank  :)

**Reply by b0661** [reply]
In #1893 ist beschrieben wie man sich die Firmware selber bauen kann. Hier https://github.com/tbnobody/OpenDTU/discussions/2134#discussioncomment-10029015 ist ein Firmware-Image für generic_esp32.

**Reply by astrakaro** [reply]
Danke für die schnelle Antwort. Doch leider funktioniert es bei nur mit der generic firmware ohne esp32
So sieht es nach dem Firmwareupdate in der Übersicht aus :
![image](https://github.com/user-attachments/assets/4a06d39a-bcf2-483e-9542-7a9d66450ac2)
Hier ist auch alles unkonfiguriert:
![image](https://github.com/user-attachments/assets/38124734-2206-4f69-9f3d-34f64772de3f)

**Reply by b0661** [reply]
Die Konfiguration der Funk-Chips hat Nichts mit Modbus zu tun. Vorgehen hier entsprechend OpenDTU Dokumentation.

Hier ein Image für generic: 
[firmware.bin.gz](https://github.com/user-attachments/files/16235601/firmware.bin.gz)

**Reply by astrakaro** [reply]
Vielen Dank!

**Comment by gmbh07** [comment]
Hallo,

könnte ich mit der Modbus Firmware und einem Smartfox eine Regelung für die Einspeisung realisieren?

**Reply by b0661** [reply]
Die momentane Modbus Firmware kann nur Daten auslesen. Es ist keine Steuerung des Hoymiles vorgesehen.

**Comment by gmbh07** [comment]
Hallo,

könnte mir jemand so eine Modbus Firmware erstellen, ich kenne mich hier leider nicht aus.

Für das Update verwende ich immer folgende Datei: opendtu-generic_esp32s3_usb.bin

Wäre dankbar für die Hilfe
